typed-racket-lib/typed-racket/rep/free-ids.rkt

```diff
@@ -0,0 +1,82 @@
+#lang racket/base
+
+(require "../utils/utils.rkt"
+         (contract-req)
+         "rep-utils.rkt"
+         "object-rep.rkt"
+         racket/match
+         racket/set
+         racket/list)
+
+(provide/cond-contract
+ [free-ids (-> Rep? (listof identifier?))]
+ [arg-deps->idx-order (-> (listof (listof identifier?))
+                          (listof exact-nonnegative-integer?))]
+ [cycle-in-arg-deps? (-> (listof (listof identifier?))
+                         (or/c #f (listof identifier?)))])
+
+
+
+;; since we use locally nameless representation for binders,
+;; any ids we see are by definition free, so we can just cruise
+;; over the type AST grabbing any identifiers we see!
+(define (free-ids rep)
+  (define ids '())
+  (let loop! ([rep rep])
+    (match rep
+      [(Path: flds (? identifier? id))
+       (for-each loop! flds)
+       (unless (member id ids free-identifier=?)
+         (set! ids (cons id ids)))]
+      [_ (Rep-for-each rep loop!)]))
+  ids)
+
+
+;; - - - - - - - - - - - - - - - - - - - - - - - - - -
+;; helper function for dependent arg id calculations
+;; - - - - - - - - - - - - - - - - - - - - - - - - - -
+;; note: in my testing (up to 12 args or so) lists
+;; seemed to be the clear winner over free-id-tables
+;; in terms of speed/memory/etc, so lists it is!
```

Comment:
Can you put the test code in a gist or something, and link to it?

Reply:
I don't have the code any more =(

Are we interested in having some on hand?

Is it unwise to assume just sticking with the list representation for these dependencies (i.e. what we parse from the syntax) is going to be faster than bothering to build free-id-tables and other data structures for this calculation? I'm just trying to imagine in practice how many args are there going to be, how large will these dependency lists actually be, etc.